keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -0,0 +1,223 @@
+- Defaulting the configs, that are generated by kubeadm or supplied by users, populates additional fields with default values. This makes it difficult to determine in retrospect, what fields were set on purpose with their default values, and what fields were simply automatically filled in by the defaulting code.
+- Providing support for validation by vendoring component specific code can lead to false-positive and missed errors. This can be caused by the deviation between the imported code by kubeadm and the actual component.
+- Migrating between config versions also requires code to be vendored in or forked from the components. This can also lead to difficult to diagnose errors and changed behaviour in unexpected ways.
+- Not distinguishing between kubeadm generated and user supplied component configs, necessitates the requirement for config migration. Kubeadm generated configs that are not modified by users directly in the config map, can be fully regenerated upon config format change. In all other cases kubeadm must convert the config to preserve user changes.
+
+These observations along with the emergence of kubeadm phases, kustomize support and work around addon operators allows us to have the following future goals.
+
+### Goals
+
+- Stop vendoring internal types and validation, defaulting and conversion code.
+- Stop defaulting component configs.
+- Outsource validation to the component binaries themselves.
+- Distinguish between user supplied and kubeadm generated and unmodified component configs.
+- Eradicate the need to convert configs
+- Use more community wide solutions, like Kustomize, kubeadm phases and addon operators
+
+### Non-Goals
+
+- Change the handling of kubeadm’s own config types in any way.
+- Cover in any way the migration of command line flags to component configs.
+- Define strategies for config map naming and backup.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### Stop Component Config defaulting
+
+Component config defaulting has a number of issues:
+- It’s difficult to tell in retrospect if a setting value was intended or simply defaulted.
+- If a default value changes a defaulted setting will stop us from using the new value.
+- Defaulting bloats the resulting config, thus making it hard for users to migrate it upon a version change.
+
+Hence, kubeadm should stop defaulting component configs and leave this to components themselves.
+
+### Delegate config validation to the components
+
+Currently, kubeadm vendors in pieces of Go code from the components to be able to perform component config validation. This is a bad practice as it bloats the kubeadm dependencies and deviation between the vendored code and the components can lead to inconsistencies.  
+To solve this, validation is going to be performed by executing a container or the binary of the component to be used with an appropriate command to verify the supplied config.  
```

Comment:
it imposes the following sequence in kubeadm:
- load user provided componentconfig
- prepull image for component
- create container and validate the user config

Reply:
With "prepull image for component" optional as some components (e.g. kubelet) can be local binaries.
I am working today on KEP for this (to be filed with WG Component Standard).

Reply:
For the containerized components, the plan is to use the chosen container runtime to spin up the container for validation, right?

Reply:
So, this makes deployment and ordering tricky now b/c in reality components should manage their apis and thus their conversions.

Reply:
@ereslibre yes, that's right.
@timothysc yes, and API machinery don't like the idea of exposing conversion to users too much. Hence, we go on the path of regenerating automatically generated stuff by kubeadm and calling users to perform manual migration if the config is user supplied.

The KEP which I was referring is #1264

Reply:
[1] we should ensure those steps are part of "preflight", so to keep at the minimum the number of dependencies/the potential source of errors during in the middle of the kubeadm workflows